alf/algorithms/functional_particle_vi_algorithm.py

```diff
@@ -0,0 +1,518 @@
+                params evaluated on sampled extra data.
+        """
+        # sample extra data
+        if isinstance(params, tuple):
+            params, extra_samples = params
+        else:
+            sample = data[-self._function_extra_bs:]
+            noise = torch.zeros_like(sample)
+            if self._function_extra_bs_sampler == 'uniform':
+                noise.uniform_(0., 1.)
+            else:
+                noise.normal_(mean=0., std=self._function_extra_bs_std)
+            extra_samples = sample + noise
+
+        num_particles = params.shape[0]
+        self._param_net.set_parameters(params)
+        aug_data = torch.cat([data, extra_samples], dim=0)
+        aug_outputs, _ = self._param_net(aug_data)  # [B+b, P, D]
+
+        outputs = aug_outputs[:data.shape[0]]  # [B, P, D]
+        outputs = outputs.transpose(0, 1)
+        outputs = outputs.view(num_particles, -1)  # [P, B * D]
+
+        density_outputs = aug_outputs[-extra_samples.shape[0]:]  # [b, P, D]
+        density_outputs = density_outputs.transpose(0, 1)
+        density_outputs = density_outputs.view(num_particles, -1)
+
+        return outputs, density_outputs
+
+    def _function_neglogprob(self, targets, outputs):
+        """
+        Function computing negative log_prob loss for function outputs.
+        Used when function_vi is True.
+
+        Args:
+            targets (torch.Tensor): target values of the training batch.
+            outputs (torch.Tensor): function outputs to evaluate the loss.
+
+        Returns:
+            negative log_prob for outputs evaluated on current training batch.
+        """
+        num_particles = outputs.shape[0]
+        # [B, D] -> [B, N, D]
+        targets = _expand_to_replica(targets, num_particles,
+                                     self._param_net.output_spec)
+        # [B, N, D] -> [N, B, D]
+        targets = targets.permute(1, 0, 2)
+        # [N, B, D] -> [N, -1]
+        targets = targets.view(num_particles, -1)
+
+        return self._loss_func(outputs, targets)
+
+    def _neglogprob(self, inputs, params):
+        """
+        Function computing negative log_prob loss for generator outputs.
+        Used when function_vi is False.
+
+        Args:
+            inputs (torch.Tensor): (data, target) of training batch.
+            params (torch.Tensor): generator outputs to evaluate the loss.
+
+        Returns:
+            negative log_prob for params evaluated on current training batch.
+        """
+        self._param_net.set_parameters(params)
+        num_particles = params.shape[0]
+        data, target = inputs
+        output, _ = self._param_net(data)  # [B, N, D]
+        # [B, d] -> [B, N, d]
+        target = _expand_to_replica(target, num_particles,
+                                    self._param_net.output_spec)
+        return self._loss_func(output, target)
+
+    def evaluate(self):
+        """Evaluate on a randomly drawn ensemble. """
```

Comment:
not randomly drawn ensemble. The ensemble is already chosen.

Reply:
Updated.